### ISSUE TITLE
the value in list of idxs is out of the index of the weight 

### ISSUE THREAD

**Issue by purinsat** [arturjordao/PruningNeuralNetworks#1]
I tried to adapt your VIPPruning file with my simple model (Alexnet) with CIFAR10 Dataset and I got that it's a very interesting approach to compressing the model but I tried to understand how to calculate the idxs list. I can't make it crystal clear of how it come . So, could you please explain more about function idxs_to_prune in VIPPruning class or just recommend me some nice source to let's me crystal clear about how the idxs come. Thank you

**Comment by arturjordao** [comment]
Dear @PurinSatthamn,

The idx_to_prune returns a list of pairs of indices, where each element in this list represents a conv. layer index and the filters to be removed in this layer.
For example, consider the following architecture: input -> conv (idx=0) -> bn -> conv (idx=1) -> act -> conv (idx=2).
The idx_to_prune will return something like : [(0, [0, 1, 10, 54]), (1, [14, 34, 3]), (2, [])].
The middle element of this list, (1, [14, 34, 3]), indicates that we need to remove the filters 
14, 34, and 3 of the convolutional layer 1. On the other hand, the last element, (2, []), indicates we can't remove any filter from layer 2.

To generate such a list the process is the following. First,  we run the proposed PLS+VIP criterion (line 160). 
At this stage, we have generated a set of importance representing the scores of each filter composing the network (self.score_layer).  Then, we find a threshold that eliminates p% (self.percentage_discard) of filters (line 166). 
Finally, we include the filters (and their corresponding layer) below this threshold into a list.  This list will be idx_to_prune.

I hope this helps you. If you have any question please don't hesitate to contact me.
Best regards,
Artur Jordão.

**Comment by purinsat** [comment]
Dear @arturjordao,

Thanks for replying, I'm understanding more about idxs now. But I have another question when I'm in the iteration I print out the idxs filter that gonna be pruned and I found that when I created the Alexnet model with this code
    Alexnet_Model = keras.models.Sequential([
        keras.layers.Conv2D(filters=96, kernel_size=(11,11), strides=(4,4), activation='relu', input_shape=(227,227,3)),
        keras.layers.BatchNormalization(),
        keras.layers.MaxPool2D(pool_size=(3,3), strides=(2,2)),
        keras.layers.Conv2D(filters=256, kernel_size=(5,5), strides=(1,1), activation='relu', padding="same"),
        keras.layers.BatchNormalization(),
        keras.layers.MaxPool2D(pool_size=(3,3), strides=(2,2)),
        keras.layers.Conv2D(filters=384, kernel_size=(3,3), strides=(1,1), activation='relu', padding="same"),
        keras.layers.BatchNormalization(),
        keras.layers.Conv2D(filters=384, kernel_size=(1,1), strides=(1,1), activation='relu', padding="same"),
        keras.layers.BatchNormalization(),
        keras.layers.Conv2D(filters=256, kernel_size=(1,1), strides=(1,1), activation='relu', padding="same"),
        keras.layers.BatchNormalization(),
        keras.layers.MaxPool2D(pool_size=(3,3), strides=(2,2)),
        keras.layers.Flatten(),
        keras.layers.Dense(4096, activation='relu'),
        keras.layers.Dropout(0.5),
        keras.layers.Dense(4096, activation='relu'),
        keras.layers.Dropout(0.5),
        keras.layers.Dense(10, activation='softmax')
    ])

The Alexnet Model set the first conv layer to have 96 filters but the return of idxs list of the first layer has the value more than 96 and I got an error of np.delete cause of the index is out of range. I can fixed it by brute force this code 
                    idxs = idxs[idxs <= 96]
                    output.append((layer_idx, idxs))
then I got that I can rebuild the model now. Then I delete that brute force code  and rerun the same code and I got no error like before the rebuild model. So, my questions are

1. Does the process to get idxs list considering the structure (the filters) of the model before returning the idxs list ? 
2. If I just re-iteration (not just set the iterations to 500 and train ) but I make 10 iterations, save the model , loading back then rerun another 10 iterations. Do like this 50 times. Will the resulting model be the same as run just one time with 500 iterations ?

Thanks. Next Step I would adapt this algorithm with the EfficientNet Architecture. 

Best regards,
Satthamnuwong Purin.

**Comment by arturjordao** [comment]
Could you provide me some input data for yielding this error? I was not able to reproduce your error.

Regarding the other questions:
1.
Yes. The method considers the number of filters in each layer. Thus, it wouldn't return out of range indices.

2.
Due to the stochastic essence of training (i.e., fine-tuning) I believe the models could be different. 
If we ignore the training phase, since our criterion is deterministic the resulting model should be the same for the scenarios you mentioned. In summary, the models can be different due to training not because of our criterion.

**Comment by purinsat** [comment]
Dear @arturjordao
Thanks for replying. I am considering about the PLS/VIP criteria. What do
you think about using these criteria to pruned the complicated architecture
like EfficientNet architecture?
I'm going to try soon but get the opinion from you first, would be great
for design the experiment.

Best regards,
Satthamnuwong Purin.

**Comment by arturjordao** [comment]
Dear @PurinSatthamn , 

I believe PLS+VIP will work well on these architectures. However, note that this repository supports only plain networks since its goal is to provide an example of our criterion. In summary, you can use the PLS+VIP to identify potential filters to be removed, but, you need to implement the prune (i.e. the removal of filters) by yourselves.

I will close this issue since there are no bug issues.

Artur Jordão.
